File: plotting/plot_investment_costs_vs_num_clusters_one_method.jl
using CSV, DataFrames, Plots

# ── Load data ────────────────────────────────────────────────────────────────
input_path = "plotting/csv_data/investment_costs_summary.csv"
df = CSV.read(input_path, DataFrame)

# ── Choose method ────────────────────────────────────────────────────────────
available_methods = unique(df.method)
println("Available methods:")
for (i, m) in enumerate(available_methods)
    println("  $i. $m")
end
print("Choose method (number or name): ")
input = strip(readline())

chosen_method = if all(isdigit, input)
    available_methods[parse(Int, input)]
else
    input
end

@assert chosen_method in available_methods "Unknown method: $chosen_method"
println("Plotting method: $chosen_method")

# ── Filter & sort ─────────────────────────────────────────────────────────────
df_filtered = sort(filter(row -> row.method == chosen_method, df), :num_clusters)

if nrow(df_filtered) == 0
    println("No rows found for method: $chosen_method")
    exit(1)
end

# ── Identify technology cost columns ─────────────────────────────────────────
tech_cols  = filter(c -> startswith(c, "cost_"), names(df_filtered))
tech_labels = replace.(tech_cols, "cost_" => "")

# ── Build matrix: (n_experiments × n_techs), scaled to billions ──────────────
x_labels    = string.(df_filtered.num_clusters)   # categorical string labels
cost_matrix = Matrix(df_filtered[:, tech_cols]) ./ 1e6   # scale to millions

# Drop technologies with all-zero costs (keeps the legend clean)
nonzero_mask  = vec(any(cost_matrix .!= 0, dims=1))
cost_matrix   = cost_matrix[:, nonzero_mask]
tech_labels   = tech_labels[nonzero_mask]

n_experiments, n_techs = size(cost_matrix)

# ── Colour palette ────────────────────────────────────────────────────────────
tech_colors = Dict(
    "Battery"       => colorant"#f4a261",
    "Coal"          => colorant"#6d6875",
    "Gas"           => colorant"#e76f51",
    "Nuclear"       => colorant"#e9c46a",
    "OCGT"          => colorant"#264653",
    "Solar"         => colorant"#FFD166",
    "Wind_Offshore" => colorant"#118ab2",
    "Wind_Onshore"  => colorant"#06d6a0",
)

# ── Build stacked bar plot manually ──────────────────────────────────────────
x_positions = 1:n_experiments
p = plot(
    size        = (900, 550),
    dpi         = 150,
    xlabel      = "Number of clusters",
    ylabel      = "Investment cost (million €)",
    title       = "Investment costs by technology\nmethod: $chosen_method",
    legend      = :topright,
    xticks      = (x_positions, x_labels),
    margin      = 8Plots.mm,
    bar_width   = 0.7,
)

bottoms = zeros(n_experiments)
for (i, tech) in enumerate(tech_labels)
    values = cost_matrix[:, i]
    bar!(
        p,
        x_positions,
        values;
        bottom    = bottoms,
        label     = tech,
        color     = get(tech_colors, tech, colorant"grey"),
        linewidth = 0,
    )
    bottoms .+= values
end

# ── Save ──────────────────────────────────────────────────────────────────────
output_dir  = "plotting/figures"
mkpath(output_dir)
safe_method = replace(chosen_method, r"[^a-zA-Z0-9_]" => "_")
output_path = joinpath(output_dir, "investment_costs_stacked_$(safe_method).png")
savefig(p, output_path)
println("Plot saved to: $output_path")

File: plotting/plot_investment_costs_vs_num_clusters_one_method.py
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib.ticker as mticker
from pathlib import Path

# -----------------------------
# Settings
# -----------------------------
csv_path   = Path("plotting/csv_data/investment_costs_summary.csv")
output_dir = Path("plotting/figures")
output_dir.mkdir(parents=True, exist_ok=True)

TECH_COLORS = {
    "Battery":       "#f4a261",
    "Coal":          "#6d6875",
    "Gas":           "#e76f51",
    "Nuclear":       "#e9c46a",
    "OCGT":          "#264653",
    "Solar":         "#FFD166",
    "Wind_Offshore": "#118ab2",
    "Wind_Onshore":  "#06d6a0",
}

# -----------------------------
# Load & prepare
# -----------------------------
df = pd.read_csv(csv_path)

available_methods = df["method"].unique()
print("Available methods:")
for i, m in enumerate(available_methods):
    print(f"  {i+1}. {m}")

raw = input("Choose method (number or name): ").strip()
chosen_method = available_methods[int(raw) - 1] if raw.isdigit() else raw
assert chosen_method in available_methods, f"Unknown method: {chosen_method}"
print(f"Plotting method: {chosen_method}")

df = df[df["method"] == chosen_method].sort_values("num_clusters")

# -----------------------------
# Identify & filter tech columns
# -----------------------------
tech_cols = [c for c in df.columns if c.startswith("cost_")]
tech_names = [c.removeprefix("cost_") for c in tech_cols]

# Drop technologies that are zero across all rows
nonzero = [(col, name) for col, name in zip(tech_cols, tech_names)
           if df[col].sum() > 0]
tech_cols, tech_names = zip(*nonzero)

# Scale to millions
tech_data = [df[col].values / 1e6 for col in tech_cols]
colors     = [TECH_COLORS.get(name, "#aaaaaa") for name in tech_names]
x          = df["num_clusters"].values

# -----------------------------
# Plot
# -----------------------------
fig, ax = plt.subplots(figsize=(10, 6))

ax.stackplot(
    x,
    *tech_data,
    labels=tech_names,
    colors=colors,
    alpha=0.85,
)

ax.set_xlabel("Number of clusters", fontsize=12)
ax.set_ylabel("Investment cost (million €)", fontsize=12)
ax.set_title(f"Investment costs by technology — {chosen_method}",
             fontsize=13, fontweight="bold")
ax.set_xticks(x)
ax.tick_params(axis="x", rotation=45)
ax.yaxis.set_major_formatter(
    mticker.FuncFormatter(lambda val, _: f"€{val:,.1f}M")
)
ax.legend(title="Technology", fontsize=10, loc="upper left")
ax.grid(True, alpha=0.3, axis="y")

plt.tight_layout()

safe_method = "".join(c if c.isalnum() else "_" for c in chosen_method)
out_path = output_dir / f"investment_costs_stacked_{safe_method}.png"
plt.savefig(out_path, dpi=150)
plt.close()

print(f"\nSaved to {out_path}")
print("\nCost breakdown (million €):")
summary = df[["num_clusters"] + list(tech_cols)].copy()
summary[list(tech_cols)] /= 1e6
summary.columns = ["num_clusters"] + list(tech_names)
print(summary.to_string(index=False))